unit/deserialize.go

```diff
@@ -43,36 +43,83 @@ var (
 	ErrLineTooLong = fmt.Errorf("line too long (max %d bytes)", SYSTEMD_LINE_MAX)
 )
 
-// Deserialize parses a systemd unit file into a list of UnitOption objects.
+// Deserialize parses a systemd unit file into a list of UnitOptions.
 func Deserialize(f io.Reader) (opts []*UnitOption, err error) {
-	lexer, optchan, errchan := newLexer(f)
+	_, options, err := DeserializeAll(f)
+	return options, err
+}
+
+// DeserializeSections deserializes into a list of UnitSections.
+func DeserializeSections(f io.Reader) ([]*UnitSection, error) {
+	sections, _, err := DeserializeAll(f)
+	return sections, err
+}
+
+// DeserializeAll deserializes into UnitSections and UnitOptions.
+func DeserializeAll(f io.Reader) ([]*UnitSection, []*UnitOption, error) {
+
+	lexer, secchan, optchan, errchan := newLexer(f)
+
 	go lexer.lex()
 
-	for opt := range optchan {
-		opts = append(opts, &(*opt))
+	sections := []*UnitSection{}
+	options := []*UnitOption{}
+
+	var done bool
+
+	for {
```

Comment:
At an high level, this logic seems problematic.
There is an implicit data-dependency across the two channels (option attached to latest section), but I think this is not enforced in the code (as channel are polled in parallel, and select can pick in random order).

I think it will be better to embed the section-aware logic directly into the lexer, then having a single result channel.

Reply:
I can do this, but think given the code flow, the dependency is ok. I'll take a look at making it a single channel though.